app/services/update_namespace_stats.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+class UpdateNamespaceStats
+
+  def initialize(course)
+    @course = course
+    update_stats
+  end
+
+  def update_stats
+    wiki_namespaces = @course.tracked_namespaces
+    stats = {}
+
+    wiki_namespaces.each do |wiki_ns|
+      @wiki = wiki(wiki_ns[:wiki])
+      namespaces = wiki_ns[:namespaces]
+      wiki_stats = {}
+      namespaces.each do |ns|
+        @namespace = ns
+        wiki_stats[@namespace] = {
+          'edited_count': edited_articles_count,
+          'new_count': new_articles_count,
+          'revision_count': revision_count,
+          'user_count': user_count,
+          'word_count': word_count,
+          'references_count': references_count,
+          'views_count': views_count
+        }
+      end
+      stats[@wiki.domain] = wiki_stats
+    end
+    
+    crs_stat = CourseStat.find_by(course_id: @course.id) || CourseStat.create(course_id: @course.id)
+    crs_stat.stats_hash['namespace_stats'] = {}
```

Comment:
Yeah, the structure is right.
Okay, if we keep a parallel structure, we'll still need to specify the WIKI_DATA_STATS and namespace specific stats for wikidata inside it?

Reply:
So, in order to be more explicit, I has used the key `namespace_stats`. 
Is it okay, if we have a parallel structure with top-level keys as wiki domains, and then something like `namespace_overview_stats` and `wiki_overview_stats` keys. Ofcourse, as of now, `wiki_overview_stats` would only exist for wikidata.

Reply:
I'm not sure what the best structure would be.

I think we want to design the data structure in a way that makes it easy to think about how the stats will be used... namely, different sets of stats become different tabs of a stats component.

The one that matches that most would be one top-level key for each tab of stats. So maybe something like this?

```
{
  'www.wikidata.org' => WIKIDATA_STATS,
  'en.wikipedia.org-namespace-0' => ENWIKI_MAINSPACE_STATS,
  'en.wikibooks.org-namespace-102' => WIKIBOOKS_COOKBOOT_STATS
}
```

At least as far as we've identified so far, there aren't any features where it would be important to have different namespaces of the same wiki collected together for stats, separately from additional namespaces for other wikis for the same course.